src/CoreWCF.Primitives/src/CoreWCF/Security/ReceiveSecurityHeader.cs

```diff
@@ -0,0 +1,2188 @@
+            if ((outOfBandToken is X509SecurityToken) && (replyToken is X509SecurityToken))
+            {
+                byte[] outOfBandCertificateThumbprint = ((X509SecurityToken)outOfBandToken).Certificate.GetCertHash();
+                byte[] replyCertificateThumbprint = ((X509SecurityToken)replyToken).Certificate.GetCertHash();
+                return (CryptoHelper.IsEqual(outOfBandCertificateThumbprint, replyCertificateThumbprint));
+            }
+            else
+            {
+                return false;
+            }
+        }
+    }
+
+    class AggregateSecurityHeaderTokenResolver : CoreWCF.IdentityModel.Tokens.AggregateTokenResolver
+    {
+        SecurityHeaderTokenResolver tokenResolver;
+
+        public AggregateSecurityHeaderTokenResolver(SecurityHeaderTokenResolver tokenResolver, ReadOnlyCollection<SecurityTokenResolver> outOfBandTokenResolvers) :
+            base(outOfBandTokenResolvers)
+        {
+            if (tokenResolver == null)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull("tokenResolver");
+
+            this.tokenResolver = tokenResolver;            
+        }
+
+        protected override bool TryResolveSecurityKeyCore(SecurityKeyIdentifierClause keyIdentifierClause, out SecurityKey key)
+        {
+            bool resolved = false;
+            key = null;
+
+            resolved = this.tokenResolver.TryResolveSecurityKey(keyIdentifierClause, false, out key);
+
+            if (!resolved)
+            {
+                resolved = base.TryResolveSecurityKeyCore(keyIdentifierClause, out key);
+            }
+
+            if (!resolved)
+            {
+                resolved = SecurityUtils.TryCreateKeyFromIntrinsicKeyClause(keyIdentifierClause, this, out key);
+            }
+
+            return resolved;
+        }
+
+        protected override bool TryResolveTokenCore(SecurityKeyIdentifier keyIdentifier, out SecurityToken token)
+        {
+            bool resolved = false;
+            token = null;
+
+            resolved = this.tokenResolver.TryResolveToken(keyIdentifier, false, false, out token);
+
+            if (!resolved)
+            {
+                resolved = base.TryResolveTokenCore(keyIdentifier, out token);
+            }
+
+            if (!resolved)
+            {
+                for (int i = 0; i < keyIdentifier.Count; ++i)
+                {
+                    if (this.TryResolveTokenFromIntrinsicKeyClause(keyIdentifier[i], out token))
+                    {
+                        resolved = true;
+                        break;
+                    }
+                }
+            }
+
+            return resolved;
+        }
+
+        bool TryResolveTokenFromIntrinsicKeyClause(SecurityKeyIdentifierClause keyIdentifierClause, out SecurityToken token)
+        {
+            throw new NotImplementedException();
+            //token = null;
```

Comment:
Why was this implementation not brought over?

Reply:
ported... only one clarification needed, tracking it separately will check.. resolving